src/EditorFeatures/Core/Implementation/Interactive/InteractiveSession.cs

```diff
@@ -0,0 +1,367 @@
+        }
+
+        internal async Task<bool> ResetAsync(InteractiveHostOptions options)
+        {
+            try
+            {
+                // Do not queue reset operation - invoke it directly.
+                // Code execution might be in progress when the user requests reset (via a reset button, or process terminating on its own).
+                // We need the execution to be interrupted and the process restarted, not wait for it to complete.
+
+                var result = await Host.ResetAsync(options).ConfigureAwait(false);
+
+                // Note: Not calling UpdatePathsNoLock here. The paths will be updated by ProcessInitialized 
+                // which is executed once the new host process finishes its initialization.
+
+                return result.Success;
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e))
+            {
+                throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        public InteractiveHostOptions GetHostOptions(bool initialize, InteractiveHostPlatform? platform)
+            => InteractiveHostOptions.CreateFromDirectory(
+                _hostDirectory,
+                initialize ? _languageInfo.InteractiveResponseFileName : null,
+                CultureInfo.CurrentUICulture,
+                 platform ?? Host.OptionsOpt?.Platform ?? InteractiveHost.DefaultPlatform);
+
+        private static RuntimeMetadataReferenceResolver CreateMetadataReferenceResolver(IMetadataService metadataService, InteractiveHostPlatformInfo platformInfo, ImmutableArray<string> searchPaths, string baseDirectory)
+        {
+            return new RuntimeMetadataReferenceResolver(
+                searchPaths,
+                baseDirectory,
+                gacFileResolver: platformInfo.HasGlobalAssemblyCache ? new GacFileResolver(preferredCulture: CultureInfo.CurrentCulture) : null,
+                platformAssemblyPaths: platformInfo.PlatformAssemblyPaths,
+                fileReferenceProvider: (path, properties) => metadataService.GetReference(path, properties));
+        }
+
+        private static SourceReferenceResolver CreateSourceReferenceResolver(ImmutableArray<string> searchPaths, string baseDirectory)
+            => new SourceFileResolver(searchPaths, baseDirectory);
+
+        public async Task SetPathsAsync(ImmutableArray<string> referenceSearchPaths, ImmutableArray<string> sourceSearchPaths, string workingDirectory)
+        {
+            try
+            {
+                var result = await Host.SetPathsAsync(referenceSearchPaths.ToArray(), sourceSearchPaths.ToArray(), workingDirectory).ConfigureAwait(false);
+                UpdatePathsNoLock(result);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e))
+            {
+                throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        private void UpdatePathsNoLock(RemoteExecutionResult result)
```

Comment:
Usually "NoLock" implies the caller is expected to take a lock, but I don't see SetPathsAsync doing that either.

Reply:
Seems like a bug.

Reply:
Fixed.